src/simple_paths.rs

```diff
@@ -0,0 +1,155 @@
+use std::iter::{from_fn, FromIterator};
+
+use indexmap::IndexSet;
+
+use crate::{
+    Direction::Outgoing,
+    EdgeType,
+    Graph,
+    graph::IndexType,
+    prelude::NodeIndex,
+};
+
+/// Returns iterator that produces all simple paths from `from` node to `to`, which contains at least `min_intermidiate_nodes` nodes
+/// and at most `max_intermidiate_nodes`, if given, limited by graph's order otherwise
+/// Simple path is path without repetitions
+/// Algorithm is adopted from https://networkx.github.io/documentation/stable/reference/algorithms/generated/networkx.algorithms.simple_paths.all_simple_paths.html
+pub fn all_simple_paths<'g, TargetIter: 'g, N, E, Ty, Ix>(graph: &'g Graph<N, E, Ty, Ix>,
+                                                      from: NodeIndex<Ix>,
+                                                      to: NodeIndex<Ix>,
+                                                      min_intermidiate_nodes: usize,
+                                                      max_intermidiate_nodes: Option<usize>) -> impl Iterator<Item=TargetIter> + 'g
+    where Ix: IndexType, Ty: EdgeType, TargetIter: FromIterator<NodeIndex<Ix>>
```

Comment:
The TargetIter type parameter should have a different name - it's not an iterator, so Iter should not be in the name.